[3 - 4단계 방탈출 예약 대기] 리원(김지윤) 미션 제출합니다.

awrion3 opened the pull request:
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr/studylogs/4347)에 셀프 체크를 작성했나요?

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.
- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
1. 한 메서드에 오직 한 단계의 들여쓰기만 한다.
2. else 예약어를 쓰지 않는다. ✔️
3. 모든 원시값과 문자열을 포장한다.
4. 한 줄에 점을 하나만 찍는다.
5. 축약해서 줄여쓰지 않는다. ✔️
6. 모든 엔티티를 작게 유지한다.
7. 인스턴스 변수를 3개 이상 가진 클래스를 쓰지 않는다.
8. 일급 콜렉션을 쓴다.
9. 게터, 세터, 프로퍼티를 쓰지 않는다.

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

이번 스텝2에서의 [커밋 내역](https://github.com/woowacourse/spring-roomescape-waiting/pull/303/files/0d95cd7783bd7c1f5d84a76cbaa056f67b0998f6..686e2b8a5aaea00eeb60dba8f05afbedeed446b2)입니다.

## 어떤 부분에 집중하여 리뷰해야 할까요?

안녕하세요 이프, 리뷰이 리원입니다! 지난 스텝1에서 말씀하신 [피드백](https://github.com/woowacourse/spring-roomescape-waiting/pull/204#pullrequestreview-2854810660)을 반영한 것과 함께, 
이번 스텝2에서는 예약 대기 관련 기능을 어떻게 추가할지 고민해볼 수 있었습니다.
현재로서는 이를 기존 예약 도메인에 상태 필드를 추가해 예약과 대기 내역을 구분하는 방식으로 구현해보게 되었습니다.
또한 대기 내역을 예약으로 전환하는 기능의 경우, 어드민이 예약 대기 내역을 직접 승인하는 방식으로 구현해보았습니다.
이번 주말에도 시간 내어 좋은 리뷰 해주셔서 감사합니다. ☺️

### 📌 코드리뷰 질문 그라운드룰
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

awrion3 committed:
init: 이전 미션 프로젝트 파일 추가

awrion3 committed:
build: JPA 의존성 추가

awrion3 committed:
feat: Theme 엔티티 매핑

awrion3 committed:
feat: ReservationTime 엔티티 매핑

awrion3 committed:
feat: Theme Repository 구현

awrion3 committed:
feat: ReservationTime Repository 구현

awrion3 committed:
feat: Member 엔티티 매핑

awrion3 committed:
feat: Member Repository 구현

awrion3 committed:
feat: Reservation 엔티티 매핑

awrion3 committed:
feat: 엔티티 간 연관관계 매핑

awrion3 committed:
env: 테스트용 프로퍼티즈 구성

awrion3 committed:
feat: Member 서비스에 기존 DAO에서 JPA 의존으로 변경

awrion3 committed:
feat: ReservationTime 서비스에 기존 DAO에서 JPA 의존으로 변경

awrion3 committed:
feat: Theme 서비스에 기존 DAO에서 JPA 의존으로 변경

awrion3 committed:
feat: Reservation 서비스에 기존 DAO에서 JPA 의존으로 변경

awrion3 committed:
feat: 내 예약 목록 조회 기능 구현

awrion3 committed:
fix: 관리자 예약 조회 필터링 오류 수정

awrion3 committed:
test: 검색 조건에 따른 예약 내역 조회 관련 테스트 추가

awrion3 committed:
refactor: 기존 Dao와 DaoImpl 제거 및 JPA 쿼리문 추가

- EOF 줄바꿈 추가 및 사용하지 않는 메서드 제거
- data.sql 데이터 일부 수정
- 일부 하드코딩된 값 상수로 변경

awrion3 committed:
refactor: 일부 메서드 구조 개선

- 메서드 분리 및 스트림 적용
- 정적 팩토리 메서드 적용
- 테스트의 상태 코드 값 상수로 대체
- 테스트의 중복 코드 부분 개선

awrion3 committed:
refactor: 설정 파일 갱신 및 엔티티 어노테이션 추가

awrion3 committed:
refactor: 서비스 메서드 구조 개선

- 기존 DAO 제거 및 예외 처리 보완
- 동등 비교 관련 어노테이션 추가
- 변수명 명료화 및 쿼리문 구조 개선

awrion3 committed:
refactor: 테스트 메서드 일부 구조 개선

- 사이즈를 비교하는 방식으로 검증 방식 개선
- 중첩 테스트 구조 방식으로 중복 부분 제거
- 테스트 메서드명 추상화

awrion3 committed:
feat: 예약 대기 요청 및 취소 기능 구현

- 예약 객체에 상태 필드 추가
- 리소스 파일 업데이트

awrion3 committed:
feat: 어드민 예약 대기 조회, 승인, 거절 기능 구현

awrion3 committed:
test: 예약 대기 기능 관련 케이스 추가 및 일부 수정

awrion3 committed:
refactor: 대기 순서 추가 및 코드 개선

- 로그인 인증 설정 값 관련 검증 추가
- 도메인 생성자 관련 설정 추가
- 상수명 명료화 및 delete 쿼리 단순화
- 대기 순서 관리 클래스 추가

awrion3 committed:
fix: 어드민 에약 대기 승인 기능 예외 처리 추가

awrion3 committed:
fix: resolve conflicts

sinb57 commented on src/main/java/roomescape/member/login/authentication/LoginAuthenticationResolver.java:
```diff
@@ -37,6 +39,12 @@ public Object resolveArgument(
+
+        if (!annotation.required()) {
+            throw new AuthenticationException(LoginExceptionMessage.AUTHENTICATION_MISSING.getMessage());
+        }
```
왜 Exception을 던지는지 의미 파악을 못했어요. required 가 false인데 왜 Exception을 발생시키나요?

awrion3 replied:
음 네, 다시 생각해보니 현재로서는 항상 로그인 인증이 필요하다는 점에서 굳이 required 속성을 구현해둘 필요가 없다고 여겨지네요. 이에 따라 required false로 지정한 경우에 대한 예외 처리도 불필요하다고 파악되어, 해당 코드 부분을 제거하는 방향으로 수정했습니다.

sinb57 commented on src/main/java/roomescape/common/exception/message/ReservationExceptionMessage.java:
```diff
@@ -2,7 +2,8 @@
 public enum ReservationExceptionMessage {
     TIME_BEFORE_NOW("당일의 과거 시간대로는 예약할 수 없습니다."),
-    DUPLICATE_RESERVATION("이미 동일한 예약이 존재합니다.");
+    DUPLICATE_RESERVATION("이미 동일한 예약이 존재합니다."),
+    OCCUPIED_RESERVATION("해당 대기 내역은 아직 예약으로 전환될 수 없습니다.");
```
팀 컨벤션마다 다르지만, 아래와 같이 선언하는 것을 선호하는 경우도 있어요.

```java
    OCCUPIED_RESERVATION("해당 대기 내역은 아직 예약으로 전환될 수 없습니다."),
    ;
```

이 경우 file change를 파악하기 수월할 때, 코드의 5번째 행처럼 `;` -> `,` 변경이 딸려오지 않아서 좋거든요.

awrion3 replied:
오 네, 이렇게 선언하는 방식도 있군요! 적용해서 수정해보겠습니다!

sinb57 commented on src/main/java/roomescape/reservation/controller/ReservationApiController.java:
```diff
@@ -3,20 +3,20 @@
 import roomescape.reservation.service.ReservationService;
 
-@RestController
+@Controller
```
왜 `@RestController`에서 `@Controller`로 변경했는지 궁금해요~

awrion3 replied:
음 네, ResponseEntity를 사용하는 경우 굳이 중복해서 RestController를 사용할 필요가 없다고 판단해 관련 부분들을 수정해보게 되었습니다!

sinb57 replied:
왜 `ResponseEntity`를 사용하는 방향으로 컨벤션을 정했는지 궁금해요~

awrion3 replied:
네, `@RestController`의 경우 기본적으로 200 ok로만 반환하게 됩니다. 따라서 `ResponseEntity`를 사용해 바디 구조와 상태 코드를 유연하게 구성하고자 했습니다.

sinb57 replied:
단순히 상태코드를 정의하기 위함이라면 `@ResponseStatus` 어노테이션을 사용하는 것도 하나의 방법인데요. 어떻게 생각하시나요??

awrion3 replied:
음 네, ResponseStatus는 특정 HTTP 상태 코드를 각 매핑 메서드에 개별적으로 지정할 수 있지만, ResponseEntity를 사용하면 상태 코드뿐 아니라 응답 바디, 필요 시 헤더 등까지 함께 조작할 수 있다는 점에서 좀 더 유연하고 일관된 응답 구성이 가능하다고 생각해서 사용하게 된 것 같습니다.

sinb57 commented on src/main/java/roomescape/reservation/domain/WaitingOrder.java:
```diff
@@ -0,0 +1,17 @@
+import java.util.concurrent.atomic.AtomicInteger;
+import org.springframework.stereotype.Component;
+
+@Component
+public class WaitingOrder {
+    private static final AtomicInteger WAITING_ORDER_VALUE = new AtomicInteger(0);
```
`@Component` 로 지정한 이유가 궁금해요~

awrion3 replied:
음 네, 해당 객체는 MyReservationService에서 생성자 주입을 통해 사용되고 있는데요. 그래서 우선 Spring 컨테이너에 등록하여 싱글턴 인스턴스로 관리하고자 했습니다.

sinb57 replied:
`GET /reservations-mine` 요청이 동시에 여럿 들어오면 어떻게 될지 고민해보시겠어요??

awrion3 replied:
아 네, 멀티 스레드 환경을 고려해서 static 선언을 붙여 전체 시스템이 대기 번호 생성 방식을 공유하도록 다시 수정했습니다!

sinb57 replied:
`공유`하는 행위가 현재 구조의 문제인데요.

GET /reservations-mine 요청이 동시에 여럿 들어오면, 어떤 문제가 발생할까요?

그리고 왜 Service나 Controller 등의 Bean 들은 상태를 지니는 것이 안티패턴일까요?

awrion3 replied:
아 네, 다시 생각해보니 공유함으로써 스레드별 대기 순서 값이 서로 영향을 줄 수 있겠네요! 이를 고려해서 대기 순번 관련 코드를 다시 수정했습니다. 그리고 컨트롤러 등도 상태를 지니는 경우, 싱글턴으로 생성된 객체이기에 만약 여러 요청이 동시에 해당 Bean에 접근하면 예상치 못한 값이 공유되거나 덮어쓸 수 있기에 좋지 않다고 이해했습니다.

sinb57 replied:
변경된 코드를 보니 WaitingOrder는 단순히 인덱스를 증가/반환 하는 용도로만 사용되는데요. WaitingOrder가 필요한지 의문이에요. List<Reservation>의 인덱스를 활용하면 그자체가 order이지 않을까요???

awrion3 replied:
오 네, 굳이 클래스로 분리할 필요가 없겠네요. MyReservationService 메서드로 옮겨 스트림 내부에서 AtomicLong을 활용해 처리하는 방식으로 수정했습니다!

sinb57 replied:
한 메서드안에서 처리되는 로직이고 동시성을 고려할 이유가 없기 때문에, AtomicLong 사용은 과하게 느껴지네요~

awrion3 replied:
네, AtomicLong을 사용하지 않는 방향으로 구조를 개선해보도록 하겠습니다!

sinb57 commented on src/main/java/roomescape/reservation/service/MyReservationService.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    private MyReservationResponse generateMyReservation(Reservation reservation) {
+        if (reservation.isWaitingStatus()) {
```
`Status`가 네이밍에 포함되어야 하는지 궁금해요~ `isWaiting`만으로도 의미전달이 되지 않을까요?

awrion3 replied:
음 네, 해당 부분은 굳이 의미를 더할 필요가 없을 수 있다는 생각이 드네요! 수정했습니다!

sinb57 commented on src/main/java/roomescape/reservation/service/MyReservationService.java:
```diff
@@ -0,0 +1,37 @@
+        this.waitingOrder = waitingOrder;
+    }
+
+    public List<MyReservationResponse> findAllMyReservationByMemberId(final Long memberId) {
```
`MemberId`가 메서드명에 명시되어야 할까요?

awrion3 replied:
아 네, 다른 메서드명과 유사하게 변경하도록 하겠습니다!

sinb57 commented on src/test/java/roomescape/reservation/domain/WaitingOrderTest.java:
```diff
@@ -0,0 +1,17 @@
+
+class WaitingOrderTest {
+
+    @DisplayName("1씩 증가된 대기 순서 값을 가져온다")
+    @Test
+    void increased_order() {
+        WaitingOrder waitingOrder = new WaitingOrder();
+
+        assertThat(waitingOrder.issueNextWaitingOrder()).isEqualTo(waitingOrder.issueNextWaitingOrder() - 1);
```
`issueNextWaitingOrder` 메서드에 대한 검증이 옳바르게 이뤄졌다고 생각하시나요?

awrion3 replied:
오 네, 값을 변수에 저장하는 방식으로 비교하는 방식으로 수정해보겠습니다!

sinb57 reviewed (changes requested):
안녕하세요 리원~ 주말에도 미션 진행하느라 고생 많으셨어요!
코멘트 확인해주시기 바래요~

awrion3 committed:
refactor: 일부 메서드 네이밍과 코드 구조 개선

- 로그인 인증 어노테이션 관련 설정 제거
- 예외 메시지 형식 수정
- 컨트롤러 어노테이션 적용으로 변경
- 일부 메서드 네이밍과 테스트 검증 방식 개선

awrion3 committed:
refactor: 열거형 수정 및 어노테이션 추가

- enum 줄바꿈 구조 개선
- 트랜잭션 관련 어노테이션 설정

awrion3 committed:
refactor: 어노테이션 추가 설정 및 코드 개선

- 트랜잭션 관련 어노테이션 설정 추가
- 테스트 메서드 정정 및 상수 설정 해제

sinb57 reviewed (changes requested):
안녕하세요 리원~ 앞서 남겼던 리뷰에 추가 코멘트 남겨두었어요~ 확인 부탁드려요!

awrion3 committed:
fix: 누락된 static 선언 추가

sinb57 reviewed (changes requested):
안녕하세요 리원~ 답변 주신 내용에 추가 코멘트 남겨두었어요! 확인 부탁드려요~

awrion3 committed:
fix: 대기 순번의 갱신 방식 수정

- 일부 테스트 케이스의 날짜 값 수정

awrion3 committed:
fix: 일부 테스트 케이스의 시간 값 수정

sinb57 reviewed (changes requested):
안녕하세요 리원~ 변경하신 부분 확인했고 추가 코멘트 하나 남겼어요~!
관련 스레드만 종료되면 PR 머지토록 할게요~

awrion3 committed:
refactor: 대기 순번 관련 구조 단순화

sinb57 reviewed (approved):
안녕하세요 리원~ 미션 진행하느라 고생 많으셨어요!!
미션에서 요구하는 학습 수준을 만족하셔서 이만 PR 머지할게요~!!
좀더 다루고 싶은 내용이나 질문이 있다면 언제든 편하게 연락주세요~
